inject/src/main/java/io/micronaut/inject/annotation/AbstractAnnotationMetadata.java

```diff
@@ -38,41 +38,52 @@
 @Internal
 abstract class AbstractAnnotationMetadata implements AnnotationMetadata {
 
-    protected final Map<String, Annotation> annotationMap;
-    protected final Map<String, Annotation> declaredAnnotationMap;
-    private Annotation[] allAnnotationArray;
-    private Annotation[] declaredAnnotationArray;
-
-    /**
-     * Constructs an instance for the given arguments.
-     *
-     * @param declaredAnnotations The declared annotations
-     * @param allAnnotations All annotations
-     */
-    protected AbstractAnnotationMetadata(@Nullable Map<String, Map<CharSequence, Object>> declaredAnnotations,
-                                         @Nullable Map<String, Map<CharSequence, Object>> allAnnotations) {
-        this.declaredAnnotationMap = declaredAnnotations != null ? new ConcurrentHashMap<>(declaredAnnotations.size()) : null;
-        this.annotationMap = allAnnotations != null ? new ConcurrentHashMap<>(allAnnotations.size()) : null;
-    }
+    private volatile Map<String, Annotation> annotationMap;
+    private volatile Map<String, Annotation> declaredAnnotationMap;
+    private volatile Annotation[] allAnnotationArray;
+    private volatile Annotation[] declaredAnnotationArray;
 
     /**
      * Constructs a default metadata.
      */
     protected AbstractAnnotationMetadata() {
-        annotationMap = new ConcurrentHashMap<>(2);
-        declaredAnnotationMap = new ConcurrentHashMap<>(2);
+    }
+
+    private Map<String, Annotation> getAnnotationMap() {
+        Map<String, Annotation> map = annotationMap;
+        if (map == null) {
+            synchronized (this) { // double check
```

Comment:
can we use a reentrant read/write lock?

Reply:
We can, considering we don't synthesize annotations anywhere this might be overkill